### PR TITLE
[ML] Allow stopping DF analytics whose config is missing

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
It is possible that the config document for a data frame
analytics job is deleted from the config index. If that is
the case the user is unable to stop a running job because
we attempt to retrieve the config and that will throw.

This commit changes that. When the request is forced,
we do not expand the requested ids based on the existing
configs but from the list of running tasks instead.

### CODE REVIEW

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java [hunk 1]
```diff
@@ -46,4 +53,21 @@ public void testHasValidLengthForId() {
         assertThat(MlStrings.hasValidLengthForId(randomAlphaOfLength(64)), is(true));
         assertThat(MlStrings.hasValidLengthForId(randomAlphaOfLength(65)), is(false));
     }
+
+    public void testFindMatching_GivenEmptyItems() {
+        assertThat(MlStrings.findMatching(new String[0], Collections.emptySet()), is(empty()));
+    }
+
+    public void testFindMatching_GivenAllPattern() {
+        assertThat(MlStrings.findMatching(new String[] {"_all"}, new HashSet<>(Arrays.asList("a", "b"))), hasItems("a", "b"));
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java:62]: IIRC `hasItems` is a partial match (i.e. it asserts that the actual collection has expected items and maybe some more).
I think `org.hamcrest.Matchers.contains` should be used instead in order to achieve full match.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java [hunk 2]
```diff
@@ -46,4 +53,21 @@ public void testHasValidLengthForId() {
...
+    }
+
+    public void testFindMatching_GivenWildcardPattern() {
+        assertThat(MlStrings.findMatching(new String[] {"*"}, new HashSet<>(Arrays.asList("a", "b"))), hasItems("a", "b"));
+    }
+
+    public void testFindMatching_GivenMixedPatterns() {
+        assertThat(MlStrings.findMatching(new String[] {"concrete", "wild-*"}, new HashSet<>(
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java:70]: As noted before, please add a test when one item matches 2 patterns.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDataFrameAnalyticsAction.java [hunk 3]
```diff
@@ -112,26 +115,51 @@ protected void doExecute(Task task, StopDataFrameAnalyticsAction.Request request
...
+        } else {
+            configProvider.getMultiple(request.getId(), request.allowNoMatch(), ActionListener.wrap(
+                configs -> matchingIdsListener.onResponse(
+                    configs.stream().map(DataFrameAnalyticsConfig::getId).collect(Collectors.toSet())),
+                matchingIdsListener::onFailure
+            ));
+        }
+    }
+
+    private Set<String> getAllStartedIds(ClusterState clusterState) {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDataFrameAnalyticsAction.java:144]: static?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlStrings.java [hunk 4]
```diff
@@ -94,4 +101,33 @@ public static String getParentField(String fieldPath) {
...
+        }
+
+        List<String> matchingItems = new ArrayList<>();
+        for (String pattern : patterns) {
+            if (items.contains(pattern))  {
+                matchingItems.add(pattern);
+            } else if (Regex.isSimpleMatchPattern(pattern)) {
+                for (String item : items) {
+                    if (Regex.simpleMatch(pattern, item)) {
+                        matchingItems.add(item);
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlStrings.java:126]: It very well can! Well spotted! Fixing.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java:62]: :+1:

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/utils/MlStringsTests.java:70]: :+1:

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlStrings.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (przemekwitek) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlStrings.java:127]: Ah, I didn't notice the "Set" type, so the code was correct. Anyway, thanks for adding a test!

**Review by przemekwitek — APPROVED** [review]
LGTM

**Comment by dimitris-athanasiou** [comment]
> Ah, I didn't notice the "Set" type, so the code was correct. Anyway, thanks for adding a test!

@przemekwitek Somehow I can't reply to your comment, so I'll reply here. The code was not correct. I switched it to use a set instead of a list to fix the problem. Well spotted once more!

**Pull request merged**